boltons/dictutils.py

```diff
@@ -964,6 +964,49 @@ def subdict(d, keep=None, drop=None):
 
     keys = set(keep) - set(drop)
 
-    return dict([(k, v) for k, v in d.items() if k in keys])
+    return type(d)([(k, v) for k, v in d.items() if k in keys])
+
+
+class FrozenDict(dict):
+    _hash = None
+
+    def updated(self, *a, **kw):
+        """Make a copy and add items from a dictionary or iterable (and/or
+        keyword arguments), overwriting values under an existing
+        key. See :meth:`dict.update` for more details.
+        """
+        data = dict(self)
+        data.update(*a, **kw)
+        return type(self)(data)
+
+    @classmethod
+    def fromkeys(cls, keys, value=None):
+        # one of the lesser known and used/useful dict methods
+        return cls(dict.fromkeys(keys, value))
+
+    def __repr__(self):
+        cn = self.__class__.__name__
+        return '%s(%s)' % (cn, dict.__repr__(self))
+
+    def __reduce_ex__(self, protocol):
+        return type(self), (dict(self),)
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash(frozenset(self.items()))
```

Comment:
this flexes up to 3x memory usage temporarily right?

I guess in python3 items() is a generator so only 2x

the other implementation used an accumulator that xor'd together hashes of keys + values

as long as you use a well behaved operator (like xor or sum-modulo-2**31) the order in which the values are accumulated together doesn't matter

on the other hand, frozenset.__hash__ is in C so is probably a lot faster; and frozenset pointed at the same keys + values is not doing copies of the underlying python objects, so the memory usage isn't really 2x/3x

Reply:
Yeah, most of that occurred to me, but I figured I'd go for brevity and something I knew would be correct. The hash is cached, that's the main thing. Later, if you or I feel like optimizing we can always get in there and measure and such.